### PR TITLE
Defines PytatoPyOpenCLArrayContext.transform_dag

### PR DESCRIPTION
**Pull request by kaushikcfd**
Similar to `transform_loopy_program` allows an interface to manipulate the DAG called `transform_dag`.

Draft because:
- [x] CI failure
  - [x] Depends on https://github.com/inducer/pytato/pull/137

### CODE REVIEW

File: arraycontext/impl/pytato/__init__.py [hunk 1]
```diff
@@ -170,6 +174,22 @@ def transform_loopy_program(self, t_unit):
                          "transform_loopy_program. Sub-classes are supposed "
                          "to implement it.")
 
+    def transform_dag(self, dag: Union["pytato.Array",
```
Comment (inducer) [inline arraycontext/impl/pytato/__init__.py:177]: Add to class docstring?

File: arraycontext/impl/pytato/__init__.py [hunk 2]
```diff
@@ -170,6 +174,22 @@ def transform_loopy_program(self, t_unit):
...
+                                       "pytato.DictOfNamedArrays"]
+                      ) -> Union["pytato.Array", "pytato.DictOfNamedArrays"]:
+        """
+        Returns a transformed version of *dag*. Sub-classes are supposed to
+        override this method to implement context-specific transformations on
+        *dag* (most likely to perform domain-specific optimizations). Every
+        :mod:`pytato` DAG that is compiled to a :mod:`pyopencl` kernel is
+        operated with this transformation.
```
Comment (inducer) [inline arraycontext/impl/pytato/__init__.py:185]:
```suggestion
        passed through this routine.
```

File: arraycontext/impl/pytato/__init__.py [hunk 3]
```diff
@@ -170,6 +174,22 @@ def transform_loopy_program(self, t_unit):
                          "transform_loopy_program. Sub-classes are supposed "
                          "to implement it.")
 
+    def transform_dag(self, dag: Union["pytato.Array",
+                                       "pytato.DictOfNamedArrays"]
```
Comment (inducer) [inline arraycontext/impl/pytato/__init__.py:178]: Would it be possible to standardize this to only `DictOfNamedArrays`?

File: arraycontext/impl/pytato/__init__.py (hunk 1 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py:177]: Done, thansk!

File: arraycontext/impl/pytato/__init__.py (hunk 2 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py]: Thanks, done!

File: arraycontext/impl/pytato/__init__.py (hunk 3 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py]: Yep, agreed. Probably easier for the derived classes.

**Comment by inducer** [comment]
Thanks!

**Pull request merged**